src/DeterministicNoiseSimulator.cpp

```diff
@@ -106,169 +122,214 @@ std::map<std::string, double> DeterministicNoiseSimulator::DeterministicSimulate
+
+    std::array<dEdge, 4> new_edges{};
+    for (short i = 0; i < 4; i++) {
+        if (firstPathEdge || i == 1) {
+            // If I am to the useDensityMatrix I cannot minimize the necessary operations anymore
+            dEdge::applyDmChangesToEdges(&originalCopy.p->e[i], nullptr);
+            new_edges[i] = applyNoiseEffects(originalCopy.p->e[i], used_qubits, true);
+            dEdge::revertDmChangesToEdges(&originalCopy.p->e[i], nullptr);
+        } else if (i == 2) {
+            // Size e[1] == e[2] (due to density matrix representation), I can skip calculating e[2]
+            new_edges[2].p = new_edges[1].p;
+            new_edges[2].w = new_edges[1].w;
+        } else {
+            dEdge::applyDmChangesToEdges(&originalCopy.p->e[i], nullptr);
+            new_edges[i] = applyNoiseEffects(originalCopy.p->e[i], used_qubits, false);
+            dEdge::revertDmChangesToEdges(&originalCopy.p->e[i], nullptr);
         }
-
-        new_edges[i] = ApplyNoiseEffects(density_op.p->e[i], op, maxDepth);
-
-        // Adding the operation to the operation table
-        noiseInsert(density_op.p->e[i], used_qubits, new_edges[i]);
     }
+    dd::Edge<dd::dNode> e = {};
+    if (std::count(used_qubits.begin(), used_qubits.end(), originalCopy.p->v)) {
+        for (auto& new_edge: new_edges) {
+            if (new_edge.w.approximatelyZero()) {
+                new_edge.w = dd::Complex::zero;
+            } else {
+                new_edge.w = dd->cn.getCached(dd::CTEntry::val(new_edge.w.r), dd::CTEntry::val(new_edge.w.i));
+            }
+        }
 
-    //if (line[density_op.p->v] > 0) {
-    if (op->actsOn(density_op.p->v)) {
-        //todo when A and D noise is active and prob is 0, I lose cached variables
         for (auto const& type: gateNoiseTypes) {
             switch (type) {
                 case 'A':
-                    ApplyAmplitudeDampingToNode(new_edges);
+                    ApplyAmplitudeDampingToNode(new_edges, (used_qubits.size() == 1) ? ampDampingProbSingleQubit : ampDampingProbMultiQubit);
                     break;
                 case 'P':
-                    ApplyPhaseFlipToNode(new_edges);
+                    ApplyPhaseFlipToNode(new_edges, (used_qubits.size() == 1) ? noiseProbSingleQubit : noiseProbMultiQubit);
                     break;
-                    //case 'B':
-                    //    applyBitFlipToNode(new_edges);
-                    //    break;
                 case 'D':
-                    ApplyDepolaritationToNode(new_edges);
+                    ApplyDepolarisationToNode(new_edges, (used_qubits.size() == 1) ? noiseProbSingleQubit : noiseProbMultiQubit);
                     break;
                 default:
                     throw std::runtime_error(std::string("Unknown gate noise type '") + type + "'");
             }
         }
+
+        for (auto& new_edge: new_edges) {
+            if (new_edge.w.approximatelyZero()) {
+                if (new_edge.w != dd::Complex::zero) {
```

Comment:
Use the new `exactlyZero` function. Please also check throughout the rest of the project for such usages.

Reply:
done